lib/templates.php

```diff
@@ -259,23 +259,62 @@ function filter_rest_wp_template_query( $args, $request ) {
 add_filter( 'rest_wp_template_query', 'filter_rest_wp_template_query', 99, 2 );
 
 /**
- * Run synchrnonization for template API requests
+ * Create the templates and template parts auto-drafts for the current theme.
+ * If the current theme is a child theme then the parent theme is synchronized as well.
+ * Runs only if current or parent theme was never loaded before or it is a newer version.
  *
- * @param mixed           $dispatch_result Dispatch result, will be used if not empty.
- * @param WP_REST_Request $request         Request used to generate the response.
- * @param string          $route           Route matched for the request.
- * @return mixed Dispatch result.
+ * To force synchronization on every load define BLOCK_THEME_DEV_MODE constant with true value.
+ *
+ * ```php
+ * define( 'BLOCK_THEME_DEV_MODE', true );
+ * ```
  */
-function gutenberg_filter_rest_wp_template_dispatch( $dispatch_result, $request, $route ) {
-	if ( null !== $dispatch_result ) {
-		return $dispatch_result;
+function gutenberg_synchronize_theme_templates_on_version_change() {
+	if ( defined( 'BLOCK_THEME_DEV_MODE' ) && BLOCK_THEME_DEV_MODE ) {
```

Comment:
I'm not sure a constant is the best way to go about it, there's a ton of use-cases where the user might edit the files without having this constant:

 - a random plugin might do it
 - the user could do it in the "theme editor"

I personally feel like relying on the file's modified date like suggested by @aristath is probably better. 
I'm also concerned about the number of issues theme authors will raise where they won't be aware of this constant.

Reply:
Good point